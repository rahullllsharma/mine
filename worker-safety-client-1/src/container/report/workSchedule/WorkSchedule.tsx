import type {
  FieldValues,
  UnpackNestedValue,
  UseFormGetValues,
  ControllerRenderProps,
} from "react-hook-form";
import { useCallback, useEffect, useState } from "react";
import { Controller, useFormContext } from "react-hook-form";

import { FieldRules } from "@/components/shared/field/FieldRules";

import { messages } from "@/locales/messages";
import {
  isDateWithinRange,
  getFormattedShortDateTime,
  getFormattedFullDateTime,
} from "@/utils/date/helper";

import FieldDateTimePicker from "@/components/shared/field/fieldDateTimePicker/FieldDateTimePicker";
import { useTenantStore } from "@/store/tenant/useTenantStore.store";
import { greaterThanDate } from "./utils";
import { ConfirmDateChangeModal } from "./ConfirmDateChangeModal";

// We should expose the types without the prefix for other components
// bonus point: if matches the description in the graphql mutation!
export const workScheduleFormInputPrefix = "workSchedule";

const FORM_INPUT_NAMES = Object.freeze({
  startDatetime: `${workScheduleFormInputPrefix}.startDatetime` as const,
  endDatetime: `${workScheduleFormInputPrefix}.endDatetime` as const,
});

// Auxilary types - useful for creating and exposing the actual WorkSchedule types
type Form = typeof FORM_INPUT_NAMES;
type WorkScheduleFormValues = UnpackNestedValue<WorkScheduleFormData>;

export type WorkScheduleFormData = {
  -readonly [k in Form[keyof Form]]: string;
};

type WorkScheduleFormDataKeys = {
  -readonly [key in keyof Form]: string;
};

export type WorkScheduleProps = {
  dateLimits: {
    projectStartDate: string;
    projectEndDate: string;
  };
  isCompleted?: boolean;
} & WorkScheduleFormDataKeys;

/**
 * Validates if the date is greater than the end date.
 * It will return true or a message.
 *
 */
const validateGreaterThanEndDateWithErrorMsg =
  (
    getFormValue: UseFormGetValues<WorkScheduleFormValues>,
    primeContractor: string
  ) =>
  (start: string) => {
    const result = greaterThanDate({
      start,
      end: getFormValue(FORM_INPUT_NAMES.endDatetime),
    });

    return (
      result ||
      messages.startDateTimeGreaterThanEnd.replace(
        "{contractor}",
        primeContractor
      )
    );
  };

/**
 * Check if a datetime is within the start and end dates for the project.
 *
 * Most likely, this won't be called because the other validation function will pickup both the data & time now.
 * However, it could be useful for edge cases.
 */
const validateIsDateInRange =
  (
    { projectStartDate, projectEndDate }: WorkScheduleProps["dateLimits"],
    primeContractor: string,
    label: "Start" | "End" = "Start"
  ) =>
  (date: string) => {
    const result = isDateWithinRange(projectStartDate, projectEndDate, date);

    return (
      result ||
      messages.dateTimeNotInProjectRange
        .replace("{contractor}", primeContractor)
        .replace("{label}", label)
        .replace("{startDatetime}", getFormattedShortDateTime(projectStartDate))
        .replace("{endDatetime}", getFormattedShortDateTime(projectEndDate))
    );
  };

type ControllerRenderPropsField =
  | ControllerRenderProps<FieldValues, typeof FORM_INPUT_NAMES.startDatetime>
  | ControllerRenderProps<FieldValues, typeof FORM_INPUT_NAMES.endDatetime>
  | undefined;

/**
 * The work schedule section, reusable.
 * Although being used on the Daily Inspection Report, this can be used anywhere.
 *
 * All dates, so `startDatetime`, `endDatetime` and project limits need to be
 * properly formatted and adjusted to the local user's timezone.
 */
export default function WorkSchedule({
  dateLimits,
  startDatetime,
  endDatetime,
  isCompleted,
}: WorkScheduleProps): JSX.Element {
  const { workPackage } = useTenantStore(state => state.getAllEntities());
  const { projectStartDate, projectEndDate } = dateLimits;
  const { clearErrors, getValues, watch, setValue } = useFormContext();
  const [showConfirmModal, setShowConfirmModal] = useState(false);
  const [revertDateChange, setRevertDateChange] = useState(false);
  const [localValue, setLocalValue] = useState<string | undefined>();
  const [localField, setLocalField] = useState<ControllerRenderPropsField>();
  const [endDateResetValue, setEndDateResetValue] = useState<
    string | undefined
  >();

  const resetEndDateOnStartDateChange = useCallback(
    function (value: string) {
      // TODO: this will at least reset the end date to match the start date
      const currentEndDateTime = getValues(FORM_INPUT_NAMES.endDatetime) || "";

      const shouldUpdateEndDateTime =
        new Date(value) > new Date(currentEndDateTime);

      if (shouldUpdateEndDateTime) {
        setEndDateResetValue(currentEndDateTime);
        setValue(FORM_INPUT_NAMES.endDatetime, value, {
          shouldValidate: false,
        });
        clearErrors();
      }
    },
    [getValues, setValue, clearErrors]
  );

  useEffect(() => {
    if (revertDateChange) {
      localField?.onChange(localValue);

      if (localField?.name === FORM_INPUT_NAMES.startDatetime) {
        setValue(FORM_INPUT_NAMES.endDatetime, endDateResetValue, {
          shouldValidate: false,
        });
        setEndDateResetValue(undefined);
      }

      setLocalValue(undefined);
      setLocalField(undefined);
      setRevertDateChange(false);
      setShowConfirmModal(false);
    }
  }, [
    revertDateChange,
    localField,
    localValue,
    resetEndDateOnStartDateChange,
    setValue,
    setEndDateResetValue,
    endDateResetValue,
  ]);

  const calcMinValForEndDatetime = isCompleted
    ? undefined
    : watch(FORM_INPUT_NAMES.startDatetime) ?? projectStartDate;

  // Config rules for pickers
  const configRulesForInput = (label: "Start" | "End") => ({
    ...FieldRules.REQUIRED,
    validate: {
      isDateInRange: validateIsDateInRange(
        {
          projectStartDate,
          projectEndDate,
        },
        workPackage.attributes.primeContractor.label,
        label
      ),
    },
  });
  const startDateTimeRules = configRulesForInput("Start");

  // Event handlers
  const handleStartDateTimeChange = (
    field: ControllerRenderProps<
      FieldValues,
      typeof FORM_INPUT_NAMES.startDatetime
    >
  ): ((e?: string | undefined) => void) | undefined => {
    return (e = "") => {
      resetEndDateOnStartDateChange(e);
      if (
        new Date(e).getDate() !==
        new Date(getValues(FORM_INPUT_NAMES.startDatetime)).getDate()
      ) {
        setLocalValue(getValues(FORM_INPUT_NAMES.startDatetime));
        setLocalField(field);
      }
      field.onChange(e);
    };
  };

  const handleEndDateTimeChange = (
    field: ControllerRenderProps<
      FieldValues,
      typeof FORM_INPUT_NAMES.endDatetime
    >
  ): ((e?: string | undefined) => void) | undefined => {
    return (e = "") => {
      if (
        new Date(e).getDate() !==
        new Date(getValues(FORM_INPUT_NAMES.endDatetime)).getDate()
      ) {
        setLocalValue(getValues(FORM_INPUT_NAMES.endDatetime));
        setLocalField(field);
      }
      field.onChange(e);
    };
  };

  // Make sure dates default are changed (set without format) on daily report reopen
  useEffect(() => {
    if (!isCompleted) {
      setValue(FORM_INPUT_NAMES.startDatetime, startDatetime);
      setValue(FORM_INPUT_NAMES.endDatetime, endDatetime);
    }
  }, [isCompleted]);

  return (
    <>
      <h3 className="text-xl font-semibold text-neutral-shade-100">
        Daily Report Details
      </h3>
      <div className="mt-4" data-testid="work-schedule-section">
        <div className="flex flex-col gap-4 p-4 bg-brand-gray-10">
          <Controller
            rules={
              isCompleted
                ? undefined
                : {
                    ...startDateTimeRules,
                    validate: {
                      greaterThanEndDate:
                        validateGreaterThanEndDateWithErrorMsg(
                          getValues as unknown as UseFormGetValues<WorkScheduleFormValues>,
                          workPackage.attributes.primeContractor.label
                        ),
                      ...startDateTimeRules.validate,
                    },
                  }
            }
            name={FORM_INPUT_NAMES.startDatetime}
            defaultValue={
              // we need to format the date for the value
              // otherwise it will display the date as we get from the payload, YYYY-MM-DDThh:ss
              isCompleted
                ? getFormattedFullDateTime(startDatetime)
                : startDatetime
            }
            render={({ field, fieldState: { error } }) => (
              <FieldDateTimePicker
                {...field}
                label={`${workPackage.attributes.primeContractor.label} Work Start Day and time`}
                error={error?.message}
                min={isCompleted ? undefined : projectStartDate}
                max={isCompleted ? undefined : projectEndDate}
                required
                readOnly={isCompleted}
                onChange={handleStartDateTimeChange(field)}
                onBlur={() => localValue && setShowConfirmModal(true)}
              />
            )}
          />
        </div>
        <div className="flex flex-col gap-4 p-4 bg-brand-gray-10">
          <Controller
            rules={isCompleted ? undefined : configRulesForInput("End")}
            name={FORM_INPUT_NAMES.endDatetime}
            defaultValue={
              // we need to format the date for the value
              // otherwise it will display the date as we get from the payload, YYYY-MM-DDThh:ss
              isCompleted ? getFormattedFullDateTime(endDatetime) : endDatetime
            }
            render={({ field, fieldState: { error } }) => (
              <FieldDateTimePicker
                {...field}
                label={`${workPackage.attributes.primeContractor.label} Work End Day and time`}
                error={error?.message}
                min={calcMinValForEndDatetime}
                max={isCompleted ? undefined : projectEndDate}
                required
                readOnly={isCompleted}
                onChange={handleEndDateTimeChange(field)}
                onBlur={() => localValue && setShowConfirmModal(true)}
              />
            )}
          />
        </div>

        {showConfirmModal && (
          <ConfirmDateChangeModal
            onConfirm={() => {
              // Note: Keep the new date changes
              clearErrors();
              setLocalValue(undefined);
              setLocalField(undefined);
              setShowConfirmModal(false);
            }}
            onCancel={() => {
              // Note: Revert the form date changes to original dates
              setRevertDateChange(true);
            }}
          />
        )}
      </div>
    </>
  );
}
