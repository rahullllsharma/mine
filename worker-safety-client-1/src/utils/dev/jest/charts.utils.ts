import type { RiskLevelByDate } from "@/components/charts/dailyRiskHeatmap/types";
import { act } from "@testing-library/react";
import { RiskLevel } from "@/components/riskBadge/RiskLevel";

export function chartRenderHelper(): () => void {
  // set up a listener so we can force the width/height
  // this is required to get the chart to render in its ResponsiveWrapper
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let listener: ((rect: any) => void) | undefined;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (window as any).ResizeObserver = class ResizeObserver {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(ls: any) {
      listener = ls;
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    observe() {}
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    disconnect() {}
  };

  function triggerResize(): void {
    act(() => {
      listener &&
        listener([
          {
            contentRect: {
              x: 0,
              y: 0,
              width: 800,
              height: 800,
              top: 100,
              bottom: 0,
              left: 100,
              right: 0,
            },
          },
        ]);
    });
  }
  return triggerResize;
}

export const sampleHeatmapDates = {
  startDate: "2022-01-29",
  endDate: "2022-02-03",
};

// consumed by the DailyRiskHeatmap
export const sampleRiskLevelByDate: RiskLevelByDate[] = [
  {
    date: "2022-01-29",
    riskLevel: RiskLevel.HIGH,
  },
  {
    date: "2022-01-30",
    riskLevel: RiskLevel.MEDIUM,
  },
  {
    date: "2022-01-31",
    riskLevel: RiskLevel.UNKNOWN,
  },
  {
    date: "2022-02-01",
    riskLevel: RiskLevel.LOW,
  },
  {
    date: "2022-02-02",
    riskLevel: RiskLevel.RECALCULATING,
  },
  {
    date: "2022-02-03",
    riskLevel: RiskLevel.UNKNOWN,
  },
];
