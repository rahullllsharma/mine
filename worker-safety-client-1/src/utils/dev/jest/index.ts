export * from "./intl.utils";
export * from "./testing.utils";
export * from "./charts.utils";
