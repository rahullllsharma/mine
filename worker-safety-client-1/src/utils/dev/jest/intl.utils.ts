/** Mock the default locale when using the Intl.DateTimeFormat class */
export function mockIntlDateTimeFormatLocale(locale: string): void {
  const { DateTimeFormat } = Intl;
  jest
    .spyOn(global.Intl, "DateTimeFormat")
    .mockImplementation((_, options) => new DateTimeFormat(locale, options));
}
