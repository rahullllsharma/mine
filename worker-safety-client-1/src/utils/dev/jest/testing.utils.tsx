/* istanbul ignore file */
import type { ReactNode } from "react";
import type { PropsWithClassName } from "@/types/Generic";
import { fireEvent, render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import { useTenantStore } from "@/store/tenant/useTenantStore.store";
import { TenantMock } from "@/store/tenant/utils/tenantMock";

export const formTemplate = (
  component: ReactNode,
  defaultValues: { [key: string]: any } = {}
) => {
  function Wrapper({ children }: PropsWithClassName): JSX.Element {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  }
  return <Wrapper>{component}</Wrapper>;
};

export const formRender = (
  component: ReactNode,
  defaultValues: { [key: string]: any } = {}
) => {
  return render(formTemplate(component, defaultValues));
};

// to overcome react-select limitation for testing, of not responding to click events
export const openSelectMenu = (role = "combobox", name?: RegExp): void => {
  const inputElement = screen.getByRole(role, { name });
  fireEvent.keyDown(inputElement, { key: "ArrowDown", code: "ArrowDown" });
};

export const mockTenantStore = (mock = TenantMock) => {
  const { setTenant } = useTenantStore.getState();
  setTenant(mock);
};
