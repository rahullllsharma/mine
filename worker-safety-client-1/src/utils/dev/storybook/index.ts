export * from "./WrapperForm.utils";
export * from "./WrapperModal.utils";
export * from "./WrapperScroll.utils";
