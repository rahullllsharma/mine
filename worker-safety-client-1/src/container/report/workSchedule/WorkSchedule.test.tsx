import { fireEvent, render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import { getFormattedFullDateTime } from "@/utils/date/helper";
import { formRender, formTemplate, mockTenantStore } from "@/utils/dev/jest";
import WorkSchedule, { workScheduleFormInputPrefix } from "./WorkSchedule";

const selectors = Object.freeze({
  ["startDatetime"]: /contractor work start day and time/gi,
  ["endDatetime"]: /contractor work end day and time/gi,
});

const date = "2022-01-09T09:03";
const prevDate = "2022-01-08T10:30";

const startDateTime = "2022-01-01T00:01";
const endDateTime = "2022-01-10T23:59";

const dateLimits = {
  projectStartDate: "2021-12-01T00:00",
  projectEndDate: "2022-02-01T23:59",
};

describe(WorkSchedule.name, () => {
  mockTenantStore();
  const mockOnSubmit = jest.fn();

  const changeValueForSelectorHelper = (selector: RegExp, value?: string) =>
    fireEvent.change(screen.getByLabelText(selector), { target: { value } });

  const TestWrapper = () => {
    const methods = useForm({
      mode: "all",
    });

    return (
      <FormProvider {...methods}>
        <form
          onSubmit={methods.handleSubmit(mockOnSubmit)}
          data-testid="work-schedule-form"
        >
          <WorkSchedule
            startDatetime={startDateTime}
            endDatetime={endDateTime}
            dateLimits={dateLimits}
          />
        </form>
      </FormProvider>
    );
  };

  beforeEach(() => {
    jest.spyOn(window.console, "error").mockImplementation(() => false);
  });

  afterEach(() => {
    mockOnSubmit.mockClear();
  });

  it("should render correctly", () => {
    const { asFragment } = formRender(
      <WorkSchedule
        startDatetime={startDateTime}
        endDatetime={endDateTime}
        dateLimits={dateLimits}
      />
    );

    screen.getByRole("heading", { name: /daily report details/i });
    expect(screen.getByLabelText(selectors["startDatetime"])).toHaveValue(
      startDateTime
    );
    expect(screen.getByLabelText(selectors["endDatetime"])).toHaveValue(
      endDateTime
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with all the inputs passed, including optional time params", () => {
    formRender(
      <WorkSchedule
        startDatetime={startDateTime}
        endDatetime={endDateTime}
        dateLimits={dateLimits}
      />
    );

    expect(screen.getByLabelText(selectors["startDatetime"])).toHaveValue(
      startDateTime
    );
    expect(screen.getByLabelText(selectors["endDatetime"])).toHaveValue(
      endDateTime
    );
  });

  it("should have ranges for the start and end dates datepickers", () => {
    formRender(
      <WorkSchedule
        startDatetime={startDateTime}
        endDatetime={endDateTime}
        dateLimits={dateLimits}
      />
    );

    expect(screen.getByLabelText(selectors["startDatetime"])).toHaveValue(
      startDateTime
    );
    expect(screen.getByLabelText(selectors["startDatetime"])).toHaveProperty(
      "min",
      dateLimits.projectStartDate
    );
    expect(screen.getByLabelText(selectors["startDatetime"])).toHaveProperty(
      "max",
      dateLimits.projectEndDate
    );
    expect(screen.getByLabelText(selectors["endDatetime"])).toHaveValue(
      endDateTime
    );
    expect(screen.getByLabelText(selectors["endDatetime"])).toHaveProperty(
      "min",
      dateLimits.projectStartDate
    );
    expect(screen.getByLabelText(selectors["endDatetime"])).toHaveProperty(
      "max",
      dateLimits.projectEndDate
    );
  });

  describe("field validations", () => {
    describe.each(Object.entries(selectors))(
      "when the required '%s' field is empty",
      (id, selector) => {
        it("should display `this is a required field` error message", async () => {
          render(<TestWrapper />);

          changeValueForSelectorHelper(selector, "");

          fireEvent.submit(screen.getByTestId("work-schedule-form"));
          expect(await screen.findByLabelText(selector)).toBeInvalid();

          // This is a crappy way to test if the error span is present for each control
          expect(
            document
              .querySelector(`#${workScheduleFormInputPrefix}\\.${id}-err`)
              ?.textContent?.trim()
          ).toEqual("This is a required field");

          expect(mockOnSubmit).not.toHaveBeenCalled();
        });
      }
    );

    describe("when trying a date that is not within the date limits for the project", () => {
      it.each`
        id                 | selectedDate          | label                    | useSameDateForDatePicker
        ${"startDatetime"} | ${"2021-05-01T00:00"} | ${"Start Date and time"} | ${undefined}
        ${"startDatetime"} | ${"2023-01-20T00:00"} | ${"Start Date and time"} | ${"endDatetime"}
        ${"endDatetime"}   | ${"2021-05-01T00:00"} | ${"End Date and time"}   | ${undefined}
        ${"endDatetime"}   | ${"2023-01-20T00:00"} | ${"End Date and time"}   | ${undefined}
      `(
        "should display a error message when $id is used with the out of range date ($date)",
        async ({
          id,
          selectedDate,
          label,
          useSameDateForDatePicker,
        }: {
          id: keyof typeof selectors;
          selectedDate: string;
          label: string;
          useSameDateForDatePicker: keyof typeof selectors;
        }) => {
          const selector = selectors[id];
          render(<TestWrapper />);
          changeValueForSelectorHelper(selectors["endDatetime"], startDateTime);
          changeValueForSelectorHelper(selectors["startDatetime"], endDateTime);

          changeValueForSelectorHelper(selector, selectedDate);

          // Sometimes we need to override the other datepicker to avoid triggering other validation rules.
          if (useSameDateForDatePicker) {
            changeValueForSelectorHelper(
              selectors[useSameDateForDatePicker],
              selectedDate
            );
          }

          fireEvent.submit(screen.getByTestId("work-schedule-form"));

          expect(await screen.findByLabelText(selector)).toBeInvalid();

          // This is a crappy way to test if the error span is present for each control
          expect(
            document
              .querySelector(`#${workScheduleFormInputPrefix}\\.${id}-err`)
              ?.textContent?.trim()
          ).toEqual(
            `Contractor Work ${label} cannot start before ${new Intl.DateTimeFormat(
              undefined,
              {
                dateStyle: "short",
                timeStyle: "short",
              }
            ).format(
              new Date(dateLimits.projectStartDate)
            )} and end after ${new Intl.DateTimeFormat(undefined, {
              dateStyle: "short",
              timeStyle: "short",
            }).format(new Date(dateLimits.projectEndDate))}`
          );

          expect(mockOnSubmit).not.toHaveBeenCalled();
        }
      );
    });

    describe("when Contractor Work Start Date is greater than Contractor Work End Date", () => {
      it("it should update the end date and time and match the start date and time", async () => {
        const selector = selectors["startDatetime"];

        render(<TestWrapper />);

        changeValueForSelectorHelper(selectors["endDatetime"], prevDate);
        changeValueForSelectorHelper(selector, date);

        expect(screen.getByLabelText(selectors["endDatetime"])).toHaveValue(
          date
        );
      });
    });
  });

  it("should submit the validated inputs", async () => {
    render(<TestWrapper />);

    changeValueForSelectorHelper(selectors["endDatetime"], date);
    changeValueForSelectorHelper(selectors["startDatetime"], date);

    fireEvent.submit(screen.getByTestId("work-schedule-form"));

    expect(await screen.findByTestId("work-schedule-form")).toHaveFormValues({
      [`${workScheduleFormInputPrefix}.startDatetime`]: date,
      [`${workScheduleFormInputPrefix}.endDatetime`]: date,
    });

    expect(mockOnSubmit).toHaveBeenCalled();
  });

  it("reopen daily reports keep date input valid", () => {
    // With an complete daily report
    const { rerender } = render(
      formTemplate(
        <WorkSchedule
          startDatetime={startDateTime}
          endDatetime={endDateTime}
          dateLimits={dateLimits}
          isCompleted={true}
        />
      )
    );

    expect(screen.getByLabelText(selectors["startDatetime"])).toHaveValue(
      getFormattedFullDateTime(startDateTime)
    );
    expect(screen.getByLabelText(selectors["endDatetime"])).toHaveValue(
      getFormattedFullDateTime(endDateTime)
    );

    // If we reopen, it should keep input date valid
    rerender(
      formTemplate(
        <WorkSchedule
          startDatetime={startDateTime}
          endDatetime={endDateTime}
          dateLimits={dateLimits}
          isCompleted={false}
        />
      )
    );
    expect(screen.getByLabelText(selectors["startDatetime"])).toHaveValue(
      startDateTime
    );
    expect(screen.getByLabelText(selectors["endDatetime"])).toHaveValue(
      endDateTime
    );
  });
});
