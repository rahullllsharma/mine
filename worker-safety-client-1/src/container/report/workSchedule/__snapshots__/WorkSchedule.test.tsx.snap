// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkSchedule should render correctly 1`] = `
<DocumentFragment>
  <h3
    class="text-xl font-semibold text-neutral-shade-100"
  >
    Daily Report Details
  </h3>
  <div
    class="mt-4"
    data-testid="work-schedule-section"
  >
    <div
      class="flex flex-col gap-4 p-4 bg-brand-gray-10"
    >
      <div>
        <label
          class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
          for="workSchedule.startDatetime"
        >
          Contractor Work Start Day and time *
        </label>
        <div
          class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"
        >
          <i
            aria-hidden="true"
            class="ci-calendar ml-2 pointer-events-none w-6 h-6 text-xl leading-none bg-white text-neutral-shade-58"
          />
          <input
            aria-errormessage=""
            aria-invalid="false"
            class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 root p-2"
            id="workSchedule.startDatetime"
            max="2022-02-01T23:59"
            min="2021-12-01T00:00"
            name="workSchedule.startDatetime"
            pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}"
            type="datetime-local"
            value="2022-01-01T00:01"
          />
        </div>
      </div>
    </div>
    <div
      class="flex flex-col gap-4 p-4 bg-brand-gray-10"
    >
      <div>
        <label
          class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
          for="workSchedule.endDatetime"
        >
          Contractor Work End Day and time *
        </label>
        <div
          class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"
        >
          <i
            aria-hidden="true"
            class="ci-calendar ml-2 pointer-events-none w-6 h-6 text-xl leading-none bg-white text-neutral-shade-58"
          />
          <input
            aria-errormessage=""
            aria-invalid="false"
            class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 root p-2"
            id="workSchedule.endDatetime"
            max="2022-02-01T23:59"
            min="2021-12-01T00:00"
            name="workSchedule.endDatetime"
            pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}"
            type="datetime-local"
            value="2022-01-10T23:59"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
